server/jobmanager/serverv1/jobmanager_server_test.go

```diff
@@ -0,0 +1,349 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serverv1_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/adalton/teleport-exercise/pkg/jobmanager"
+	"github.com/adalton/teleport-exercise/pkg/jobmanager/jobmanagertest"
+	"github.com/adalton/teleport-exercise/server/jobmanager/serverv1"
+	"github.com/adalton/teleport-exercise/server/jobmanager/serverv1/testserverv1"
+	"github.com/adalton/teleport-exercise/service/jobmanager/jobmanagerv1"
+	"github.com/adalton/teleport-exercise/util/grpcutil"
+	"github.com/google/uuid"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_jobmanagerServer_Start_NoUserID(t *testing.T) {
+	jobManager := jobmanager.NewManagerDetailed(jobmanagertest.NewMockJob, nil)
+	server := serverv1.NewJobManagerServerDetailed(jobManager)
+
+	_, err := server.Start(context.Background(), &jobmanagerv1.JobCreationRequest{})
+
+	assert.NotNil(t, err)
+}
+
+func Test_jobmanagerServer_Start_WithUserID(t *testing.T) {
+	const (
+		jobName     = "myJob"
+		programPath = "/bin/ls"
+	)
+	args := []string{"-l", "/"}
+
+	jobManager := jobmanager.NewManagerDetailed(jobmanagertest.NewMockJob, nil)
+	server := serverv1.NewJobManagerServerDetailed(jobManager)
+
+	ctx := grpcutil.AttachUserIDToContext(context.Background(), "user1")
+
+	job, err := server.Start(ctx, &jobmanagerv1.JobCreationRequest{
+		Name:        jobName,
+		ProgramPath: programPath,
+		Arguments:   args,
+	})
+
+	_, parseErr := uuid.Parse(job.Id.Id)
+
+	assert.Nil(t, err)
+	assert.Nil(t, parseErr)
+	assert.Equal(t, jobName, job.Name)
+}
+
+func Test_jobmanagerServer_Start_NameExists(t *testing.T) {
+	const (
+		jobName     = "myJob"
+		programPath = "/bin/ls"
+	)
+	args := []string{"-l", "/"}
+
+	jobManager := jobmanager.NewManagerDetailed(jobmanagertest.NewMockJob, nil)
+	server := serverv1.NewJobManagerServerDetailed(jobManager)
+
+	ctx := grpcutil.AttachUserIDToContext(context.Background(), "user1")
+
+	_, _ = server.Start(ctx, &jobmanagerv1.JobCreationRequest{
+		Name:        jobName,
+		ProgramPath: programPath,
+		Arguments:   args,
+	})
+
```

Comment:
```suggestion
	_, err = server.Start(ctx, &jobmanagerv1.JobCreationRequest{
		Name:        jobName,
		ProgramPath: programPath,
		Arguments:   args,
	})
	assert.NoError(t, err)

```

Reply:
This is a bit of a philosophical point.  The choice to ignore this error, and others, in unit tests is intentional, and I believe is in line with UT best practices.  There are other tests that check the error status under the same condition and assert that under those conditions the function doesn't return an error.  The behavior of the API should be reproducible, so calling the same API again under the same condition should exhibit the same behavior.  Asserting here distracts from the use case that this test is intended to validate.

Some would argue that any given unit test should have exactly one assert.  I don't take it that far, but I do try to limit the asserts in my UTs to the specific use cases that the test is looking to validate.

If the team has a different take on when/where to have assertions in UTs I'm happy to add them, but I thought it'd be worthwhile to explain my rationale for not including them.

Reply:
I added a bunch of comments like this, so I'll elaborate why I ask for it:

- Team practice (we tend to not let errors slide, even in tests)
- Having Start pass is important for this test - in other languages you'd get an early exit via exception, in Go you won't. You don't need to go overboard for asserts that are better covered in other tests, but an `err != nil` is fine.
- It's annoying when tests swallow errors and fail mysteriously in some further line. I've lost a few minutes of my life tracking stuff like this.
- Finally, less cognitive load: as a reviewer/reader I don't have to try and figure out _why_ this wasn't handled. If we end up handling a few impossible errors because of this, that's OK.

I get the philosophical aspects of it, but that's where I ended up (for now, at least :)).

Reply:
I would argue that ignoring errors and assuming things are properly tested and verified by another test is a bit of a footgun. What if that other test which was checking the errors gets removed or has a bug?